// response/index.js

var Html   = require('html');
var Render = require('render');
var Utils  = require('utilities');
var Errors = require('errors');
var Data   = require('data');

exports.AuthenticationErrorResponse = function() {
  var html     = Errors.Message.Authentication;
  var response = Html.CreateServerlessHtmlResponse(200, html);
  return response;
}

 //TODO AjaxDiv for Render.TopSellersJson() & Render.ManufacturersData()
exports.GuestHome = function(sid, host, ts, next) {
  Render.Guest(function(serr, gresp) {
    if (serr) next(serr, null);
    else {
      Render.TopSellersJson(sid, host, function(serr, tresp) {
        if (serr) next(serr, null);
        else {
          Render.ManufacturersData(sid, host, function(serr, mresp) {
            if (serr) next(serr, null);
            else {
              var tmdiv = Html.TopManufacturersAjaxDiv(sid, false);
              var tcdiv = Html.TopCategoriesAjaxDiv   (sid, true);
              var thtml = Render.JsonResult(tresp, "top_sellers", 
                                            800, 20, 30, 20, 40);
              var html  = Html.Header(sid, host, ts) +
                          Html.IncludeFavicon() +
                          gresp.body + "<p/>" +
                          tmdiv + tcdiv +
                          thtml + "<p/>" +
                          mresp.body + "<p/>" +
                          Html.Footer();
              var response = Html.CreateServerlessHtmlResponse(200, html);
              next(null, response);
            }
          });
        }
      });
    }
  });
}

exports.MemberHome = function(sid, host, ts, username, next) {
  var tmdiv = Html.TopManufacturersAjaxDiv(sid, false);
  var tcdiv = Html.TopCategoriesAjaxDiv   (sid, true);
  var rediv = Html.RecommendedAjaxDiv     (sid, username);
  var rvdiv = Html.RecentlyViewedAjaxDiv  (sid, username);
  var cadiv = Html.CartAjaxDiv            (sid, username);
  var fadiv = Html.FavoritesAjaxDiv       (sid, username);
  var html  = Html.Header(sid, host, ts) +
              Html.IncludeFavicon() +
              Html.MemberInfoAjaxDiv(sid, username) +
              tmdiv + tcdiv + rediv + rvdiv + cadiv + fadiv +
              Html.Footer();
  var response = Html.CreateServerlessHtmlResponse(200, html);
  next(null, response);
}

exports.PurchasesHome = function(sid, host, ts, username, next) {
  Render.Purchases(sid, username, function(serr, phtml) {
    if (serr) next(serr, null);
    else {
      var tmdiv = Html.TopManufacturersAjaxDiv(sid, true);
      var tcdiv = Html.TopCategoriesAjaxDiv   (sid, true);
      var fadiv = Html.FavoritesAjaxDiv       (sid, username);
      var rediv = Html.RecommendedAjaxDiv     (sid, username);
      var html  = Html.Header(sid, host, ts) +
                  Html.MemberInfoAjaxDiv(sid, username) +
                  tmdiv + phtml.body + "<p/>" + tcdiv + fadiv + rediv +
                  Html.Footer();
      var response = Html.CreateServerlessHtmlResponse(200, html);
      next(null, response);
    }
  });
}

exports.ManufacturersHome = function(sid, host, ts, username, next) {
  Render.ManufacturersData(sid, host, function(serr, xhtml) {
    if (serr) next(serr, html);
    else {
      var tmdiv = Html.TopManufacturersAjaxDiv(sid, true);
      var rvdiv = username ? Html.RecentlyViewedAjaxDiv(sid, username) : "";
      var fadiv = username ? Html.FavoritesAjaxDiv     (sid, username) : "";
      var html  = Html.Header(sid, host, ts) +
                  tmdiv + xhtml.body + "<p/>" + rvdiv  + fadiv  +
                  Html.Footer();
      var response = Html.CreateServerlessHtmlResponse(200, html);
      next(null, response);
    }
  });
}

exports.CategoriesHome = function(sid, host, ts, username, next) {
  Render.CategoriesData(sid, host, function(serr, chtml) {
    if (serr) next(serr, html);
    else {
      var tcdiv = Html.TopCategoriesAjaxDiv(sid, true);
      var rvdiv = username ? Html.RecentlyViewedAjaxDiv(sid, username) : "";
      var fadiv = username ? Html.FavoritesAjaxDiv     (sid, username) : "";
      var html  = Html.Header(sid, host, ts) +
                  tcdiv + chtml.body + "<p/>" + rvdiv + fadiv +
                  Html.Footer();
      var response = Html.CreateServerlessHtmlResponse(200, html);
      next(null, response);
    }
  });
}

exports.BuyCart = function(sid, host, ts, username, items, next) {
  Render.BuyCart(sid, username, items, function(serr, cahtml) {
    if (serr) next(serr, null);
    else {
      var tmdiv = Html.TopManufacturersAjaxDiv(sid, false);
      var tcdiv = Html.TopCategoriesAjaxDiv   (sid, true);
      var fadiv = Html.FavoritesAjaxDiv       (sid, username);
      var rediv = Html.RecommendedAjaxDiv     (sid, username);
      var html  = Html.Header(sid, host, ts) +
                  Html.MemberInfoAjaxDiv(sid, username) +
                  tmdiv + cahtml.body + "<p/>" + tcdiv + fadiv + rediv +
                  Html.Footer();
      var response = Html.CreateServerlessHtmlResponse(200, html);
      next(null, response);
    }
  });
}

exports.ShowProduct = function(sid, host, ts, username, pid, pdata, touch,
                               next) {
  Render.Product(sid, username, pid, pdata, touch, function(serr, phtml) {
    if (serr) next(serr, null);
    else {
      var tmdiv = Html.TopManufacturersAjaxDiv(sid, true);
      var cadiv = Html.CartAjaxDiv(sid, username);
      var fadiv = Html.FavoritesAjaxDiv(sid, username);
      var rediv = Html.RecommendedAjaxDiv(sid, username);
      var html  = Html.Header(sid, host, ts) +
                  Html.Line("PRODUCT INFO:") +
                  Html.Line("SESSION: " + sid) +
                  tmdiv + phtml.body + "<p/>" + cadiv + fadiv + rediv +
                  Html.Footer();
      var response = Html.CreateServerlessHtmlResponse(200, html);
      next(null, response);
    }
  });
}

exports.Category = function(sid, host, ts, cname, cdata, next) {
  Render.Category(sid, host, cname, cdata, function(serr, chtml) {
    if (serr) next(serr, null);
    else {
      var html = Html.Header(sid, host, ts) +
                 Html.Line("CATEGORY:" + cname) +
                 Html.Line("SESSION: " + sid) + "<p/>" +
                 Render.CreateDuoCategoryDiv(sid, host, cname, 1) + "<p/>" +
                 chtml.body +
                 Html.Footer();
      var response = Html.CreateServerlessHtmlResponse(200, html);
      next(null, response);
    }
  });
}

exports.Manufacturer = function(sid, host, ts, xname, xdata, next) {
  Render.Manufacturer(sid, host, xname, xdata, function(serr, mhtml) {
    if (serr) next(serr, null);
    else {
      var html = Html.Header(sid, host, ts) +
                 Html.Line("BRAND INFO:" + xname) +
                 Html.Line("SESSION: " + sid) + "<p/>" +
                 Render.CreateDuoManufacturerDiv(sid, host, xname, 1) + "<p/>" +
                 mhtml.body +
                 Html.Footer();
      var response = Html.CreateServerlessHtmlResponse(200, html);
      next(null, response);
    }
  });
}
