// html/index.js

var Render = require('render');
var Utils  = require('utilities');

var CustomerName = process.env['CUSTOMER_NAME'];
if (!CustomerName) {
  console.error('SET ENVIRONMENT VARIABLE: "CUSTOMER_NAME"');
  process.exit(-1);
}

function create_html_response(code, body) {
  var response = {statusCode : code,
                  headers    : {"Access-Control-Allow-Origin" : "*"},
                  body       : body};
  return response;
}

exports.CreateServerlessHtmlResponse = function(code, html) {
  var response = create_html_response(code, html);
  response.headers["Content-Type"] = "text/html";
  return response;
}

exports.CreateServerlessJavascriptResponse = function(code, js) {
  var response = create_html_response(code, js);
  response.headers["Content-Type"]  = "application/javascript";
  response.headers["Cache-Control"] = "public, max-age=31536000";
  return response;
}

exports.CreateServerlessJsonResponse = function(code, json) {
  var response = create_html_response(code, JSON.stringify(json));
  response.headers["Content-Type"] = "application/json";
  return response;
}

exports.Tab = function(cnt) {
  if (!cnt) cnt = 0;
  var tabs = "";
  for (var i = 0; i < cnt; i++) tabs += "&nbsp;&nbsp;&nbsp;&nbsp;";
  return tabs;
}

exports.Line = function(txt, tabcnt) {
  if (!tabcnt) tabcnt = 0;
  var tabs = exports.Tab(tabcnt);
  return tabs + txt + "<br/>\n";
}

var AmazonDnsName  = "5ref90y7aa.execute-api.us-east-1.amazonaws.com";
var Ec2WestName    = "west-ec2.kuhiro.com";
var KuhiroWestName = "west.kuhiro.com";
var KuhiroEastName = "east.kuhiro.com";

function create_east_west_form(isaw, isa, ise, isw) {
  var ltxt, rtxt, lval, rval, t;
  if (isa) {
    ltxt = "KUHIRO";
    rtxt = "EC2";
    lval = "http://"  + Ec2WestName   + "/";
    rval = "https://" + AmazonDnsName + "/";
    t    = isaw ? "K2A" : "A2K";
  } else {
    ltxt = "WEST";
    rtxt = "EAST";
    lval = "http://" + KuhiroWestName + "/";
    rval = "http://" + KuhiroEastName + "/";
    t    = "";
  }
  var cfunc = "client_change_domain(this.value, '" + CustomerName + "', '" +
               t + "');";
  var html =
    '<input type="radio" ' + (isw ? ' checked="checked"' : '') + 
    ' onclick="' + cfunc + '" value="' + lval + '" > ' + ltxt +
    '<input type="radio" ' + (ise ? ' checked="checked"' : '') +
    ' onclick="' + cfunc + '" value="' + rval + '" > ' + rtxt;
  return html;
}

function get_body_background_color(ise, isw) {
  if      (ise) return '#FFEEEF';
  else if (isw) return '#ECFEEF';
  else          return '#FFFFFF';
}

exports.Header = function(sid, host, ts) {
  var plink   = sid ? exports.PurchasesLink(sid, ts) : "";
  var iske    = (host === KuhiroEastName);
  var iskw    = (host === KuhiroWestName);
  var isk     = (iske || iskw);
  var isae    = (host === AmazonDnsName);
  var isaw    = (host === Ec2WestName);
  var isa     = (isae || isaw);
  var ise     = (iske || isae);
  var isw     = (iskw || isaw);
  var ewlink  = "";
  if (isk || isa) {
    ewlink = create_east_west_form(isaw, isa, ise, isw);
  }
  var bgcol   = get_body_background_color(ise, isw);
  var tscript = null;
  if (ts) tscript = '<script>client_record_start_time(' + ts + ');</script>';
  var html    = '<html><head><script src="./js_helper"></script></head>' +
                '<body style="background-color:' + bgcol + '">' +
                '<table border="0" width="800">' + 
                '<tr><td style="width:450;height:30">' +
                exports.HomeLink(sid)          + exports.Tab(1) +
                exports.CategoriesLink(sid)    + exports.Tab(1) +
                exports.ManufacturersLink(sid) + exports.Tab(1) +
                plink +
                '</td><td style="width:350;height:30;text-align:right">' +
                '<span id="page_load_duration" ' +
                ' style="width:150;padding: 0px 20px 0px 0px"></span>' +
                ewlink +
                '</td></tr></table>' +
                tscript +
                '<br/>';
  return html;
}

exports.Footer = function(sid) {
  var html  = "</body></html>";
  return html;
}

function build_anchor(url, search, txt) {
  var js = "client_request('" + url + "','" + search + "');";
  return '<a onclick="' + js + '">' + txt + '</a>';
}
exports.HomeLink = function(sid) {
  var url    = './home';
  var search = sid ? 'session=' + sid : '';
  return build_anchor(url, search, 'Home');
}

exports.CategoriesLink = function(sid) {
  var url    = './show_categories';
  var search = sid ? 'session=' + sid : '';
  return build_anchor(url, search, 'Categories');
}

exports.ManufacturersLink = function(sid) {
  var url    = './show_manufacturers';
  var search = sid ? 'session=' + sid : '';
  return build_anchor(url, search, 'Brands');
}

exports.PurchasesLink = function(sid) {
  var url    = './show_purchases';
  var search = sid ? 'session=' + sid : '';
  return build_anchor(url, search, 'Purchases');
}

exports.ManufacturerLink = function(sid, xname) {
  var url    = './show_manufacturer';
  var sargs  = sid ? '&session=' + sid : '';
  var search = 'name=' + xname + sargs;
  return build_anchor(url, search, xname);
}

exports.CategoryLink = function(sid, cname) {
  var url    = './show_category';
  var sargs  = sid ? '&session=' + sid : '';
  var search = 'name=' + cname + sargs;
  return build_anchor(url, search, cname);
}

exports.ProductLink = function(sid, pid, pname) {
  var url    = './show_product_detail';
  var sargs  = sid ? '&session=' + sid : '';
  var search = 'product_id=' + pid + sargs;
  return build_anchor(url, search, pname);
}

exports.BuyProductLink = function(sid, pid, txt) {
  var url    = './buy_product';
  var sargs  = sid ? '&session=' + sid : '';
  var search = 'product_id=' + pid + '&add=true' + sargs;
  return build_anchor(url, search, txt);
}

exports.EditCartLink = function(sid, pid, txt) {
  var url    = './edit_cart';
  var sargs  = sid ? '&session=' + sid : '';
  var search = 'product_id=' + pid + '&add=true' + sargs;
  return build_anchor(url, search, txt);
}

exports.AddFavoriteLink = function(sid, pid, txt) {
  var url    = './edit_favorite';
  var sargs  = sid ? '&session=' + sid : '';
  var search = 'product_id=' + pid + '&add=true' + sargs;
  return build_anchor(url, search, txt);
}

exports.BuyCartLink = function(sid, txt) {
  var url    = './buy_cart';
  var search = sid ? 'session=' + sid : '';
  return build_anchor(url, search, txt);
}

exports.IncludeFavicon = function() {
  return '<link rel="shortcut icon" href="http://www.google.com/favicon.ico" type="image/x-icon">';
}

exports.IncludeAjaxScripts = function() {
  return '<script src="./js_helper"></script>\n';
}

function add_xml_json_request(url, id, w, th, ph, ch, clm, adiv, ndiv) {
  var html =
    '<script>\n' +
    '  client_add_xml_json_request( "' + url  + '",' +
                                   '"' + id   + '",' +
                                         w    + ','  +
                                         th   + ','  +
                                         ph   + ','  +
                                         ch   + ','  +
                                         clm  + ','  +
                                         adiv + ','  +
                                         ndiv + ');' +
    '</script>\n';
  return html;
}

var DivWidth          = 800;
var DefaultTextHeight = 20;
var ChildHeight       = 20;
var ChildImageHeight  = 30;
var ChildLeftMargin   = 20;

exports.MemberInfoAjaxDiv = function(sid, username) {
  var url = './member_info?username=' + username;
  if (sid) url += '&session=' + sid;
  return add_xml_json_request(url, "member",
                              DivWidth, 50, 0, ChildHeight,
                              ChildLeftMargin, true, 0);
}

exports.RecommendedAjaxDiv = function(sid, username) {
  var url = './show_recommended?username=' + username;
  if (sid) url += '&session=' + sid;
  return add_xml_json_request(url, "recommended",
                              DivWidth, DefaultTextHeight, 90, ChildHeight,
                              ChildLeftMargin, true, 4);
}

exports.RecentlyViewedAjaxDiv = function(sid, username) {
  var url = './show_recent?username=' + username;
  if (sid) url += '&session=' + sid;
  return add_xml_json_request(url, "recently_viewed",
                              DivWidth, DefaultTextHeight, 70, ChildHeight,
                              ChildLeftMargin, true, 3);
}

exports.TopManufacturersAjaxDiv = function(sid, mini) {
  var url = './top_manufacturers?mini=' + mini;
  if (sid) url += '&session=' + sid;
  return add_xml_json_request(url, "top_manufacturers",
                              DivWidth, DefaultTextHeight, 70, ChildImageHeight,
                              ChildLeftMargin, true, 2);
}

exports.TopCategoriesAjaxDiv = function(sid, mini) {
  var url = './top_categories?mini=' + mini;
  if (sid) url += '&session=' + sid;
  return add_xml_json_request(url, "top_categories",
                              DivWidth, DefaultTextHeight, 70, ChildImageHeight,
                              ChildLeftMargin, true, 2);
}

exports.FavoritesAjaxDiv = function(sid, username) {
  var url = './show_favorites?username=' + username;
  if (sid) url += '&session=' + sid;
  return add_xml_json_request(url, "favorites",
                              DivWidth, DefaultTextHeight, 90, ChildHeight,
                              ChildLeftMargin, true, 4);
}

exports.CartAjaxDiv = function(sid, username) {
  var url = './show_cart?username=' + username;
  if (sid) url += '&session=' + sid;
  return add_xml_json_request(url, "cart",
                              DivWidth, 40, 90, ChildHeight,
                              ChildLeftMargin, true, 4);
}

exports.AddProductSummaryScript = function(sid, host, json, tname, cnt, pid) {
  var uri    = 'show_product_summary?product_id=' + pid;
  if (sid) uri += '&session=' + sid;
  var url    = exports.CreateLoadBalancedUrl(host, uri);
  var id     = tname + "-" + cnt + "-text";
  var script = {id : id, url : url};
  json.script.push(script);
}

var MultipleDomainLoadBalancing = false;

var MaxNumDomains = 4;
var CurrDomainNum = 0;
function get_next_load_balanced_host(host) {
  CurrDomainNum++;
  var dots = host.split(".");
  var sub  = dots[0];
  var nh   = 'http://' + sub + '-' + CurrDomainNum + '.kuhiro.com';
  if (CurrDomainNum === MaxNumDomains) CurrDomainNum = 0;
  return nh;
}

exports.CreateLoadBalancedUrl = function(host, uri) {
  if (MultipleDomainLoadBalancing) {
    var nh  = get_next_load_balanced_host(host);
    var url = nh + uri + '&customer=' + CustomerName;
    return url;
  } else {
    var url = "./" + uri;
    return url;
  }
}
